crates/gitbutler-branch-actions/src/virtual.rs

```diff
@@ -860,6 +848,104 @@ pub fn commit(
     Ok(commit_oid)
 }
 
+fn join_output<'a>(stdout: &'a str, stderr: &'a str) -> Cow<'a, str> {
+    let stdout = stdout.trim();
+    if stdout.is_empty() {
+        stderr.trim().into()
+    } else {
+        stdout.into()
+    }
+}
+
+fn run_message_hook(ctx: &CommandContext, mut message: String) -> Result<String> {
+    let hook_result = git2_hooks::hooks_commit_msg(ctx.repo(), Some(&["../.husky"]), &mut message)
+        .context("failed to run hook")
+        .context(Code::CommitHookFailed)?;
+
+    if let HookResult::RunNotSuccessful { stdout, stderr, .. } = &hook_result {
+        return Err(
+            anyhow!("commit-msg hook rejected: {}", join_output(stdout, stderr))
+                .context(Code::CommitHookFailed),
+        );
+    }
+    Ok(message)
+}
+
+fn run_pre_commit_hook(ctx: &CommandContext) -> Result<()> {
+    let hook_result = git2_hooks::hooks_pre_commit(ctx.repo(), Some(&["../.husky"]))
+        .context("failed to run hook")
+        .context(Code::CommitHookFailed)?;
+
+    if let HookResult::RunNotSuccessful { stdout, stderr, .. } = &hook_result {
+        return Err(
+            anyhow!("commit hook rejected: {}", join_output(stdout, stderr))
+                .context(Code::CommitHookFailed),
+        );
+    }
+    Ok(())
+}
+
+pub(crate) fn unstage_all(ctx: &CommandContext) -> Result<()> {
+    let repo = ctx.repo();
+    // Get the HEAD commit (current commit)
+    let head_commit = repo.head()?.peel_to_commit()?;
+    // Reset the index to match the HEAD commit
+    repo.reset(head_commit.as_object(), ResetType::Mixed, None)?;
+    Ok(())
+}
+
+fn diff_workdir_to_index<'a>(repo: &'a Repository, path: &PathBuf) -> Result<git2::Diff<'a>> {
+    let index = repo.index()?;
+    let mut diff_opts = git2::DiffOptions::new();
+    diff_opts
+        .recurse_untracked_dirs(true)
+        .include_untracked(true)
+        .show_binary(true)
+        .show_untracked_content(true)
+        .ignore_submodules(true)
+        .context_lines(3)
+        .pathspec(path);
+    Ok(repo.diff_index_to_workdir(Some(&index), Some(&mut diff_opts))?)
+}
+
+fn stage_file(ctx: &CommandContext, path: &PathBuf, hunks: &Vec<VirtualBranchHunk>) -> Result<()> {
+    let repo = ctx.repo();
+    let mut index = repo.index()?;
+    if hunks.iter().any(|h| h.change_type == ChangeType::Untracked) {
+        index.add_path(path)?;
+        index.write()?;
+        return Ok(());
+    }
+
+    let mut apply_opts = ApplyOptions::new();
+    apply_opts.hunk_callback(|cb_hunk| {
+        cb_hunk.map_or(false, |cb_hunk| {
+            for hunk in hunks {
+                if hunk.start == cb_hunk.new_start()
+                    && hunk.end == cb_hunk.new_start() + cb_hunk.new_lines()
+                {
+                    return true;
+                }
+            }
+            false
+        })
+    });
+
+    let diff = diff_workdir_to_index(repo, path)?;
+    repo.apply(&diff, ApplyLocation::Index, Some(&mut apply_opts))?;
```

Comment:
`gitoxide` still can't apply patches, but once it can this can be written so that it will write the `.git/index` only once.